File: apps/frontend/src/lib/api/pipeline.ts
/**
 * Pipeline API
 */

import apiClient from "./client";

export interface StageProgress {
  stage: string;
  status: string;
  progress_percent?: number;
  message?: string;
  started_at?: string;
  completed_at?: string;
  duration_ms?: number;
  summary?: Record<string, any>;
}

export interface PipelineStatus {
  run_id: string;
  project_id: string;
  status: string;
  current_stage?: string;
  progress_percent: number;
  stages: StageProgress[];
  started_at?: string;
  completed_at?: string;
  error_message?: string;
}

export interface PipelineRun {
  id: string;
  project_id: string;
  ifc_file_id?: string;
  status: string;
  current_stage?: string;
  progress_percent: number;
  configuration: Record<string, any>;
  started_at?: string;
  completed_at?: string;
  created_at: string;
}

export interface PipelineRunConfig {
  ifc_file_id?: string;
  loin_config_id?: string;
  options?: Record<string, any>;
}

export const pipelineApi = {
  start: async (projectId: string, config?: PipelineRunConfig): Promise<PipelineStatus> => {
    const { data } = await apiClient.post(`/projects/${projectId}/pipeline/run`, config ?? null);
    return data;
  },

  getStatus: async (projectId: string, runId?: string): Promise<PipelineStatus> => {
    const params = runId ? { run_id: runId } : {};
    const { data } = await apiClient.get(`/projects/${projectId}/pipeline/status`, { params });
    return data;
  },

  getHistory: async (projectId: string, limit = 20): Promise<{ items: PipelineRun[]; total: number }> => {
    const { data } = await apiClient.get(`/projects/${projectId}/pipeline/history`, {
      params: { limit },
    });
    return data;
  },

  cancel: async (projectId: string, runId?: string): Promise<void> => {
    const params = runId ? { run_id: runId } : {};
    await apiClient.post(`/projects/${projectId}/pipeline/cancel`, null, { params });
  },

  getResult: async (projectId: string, runId: string): Promise<any> => {
    const { data } = await apiClient.get(`/projects/${projectId}/pipeline/result/${runId}`);
    return data;
  },
};

export default pipelineApi;

File: apps/backend/src/api/v1/schemas/pipeline.py
"""Pipeline schemas."""

from datetime import datetime
from typing import Optional, Any
from uuid import UUID

from pydantic import BaseModel, Field


class PipelineRunCreate(BaseModel):
    """Schema for starting a pipeline run."""

    ifc_file_id: Optional[UUID] = None  # Optional - uses first uploaded file if not specified
    loin_config_id: Optional[UUID] = None
    options: dict[str, Any] = Field(default_factory=dict)


class StageProgress(BaseModel):
    """Schema for stage progress."""

    stage: str  # parse, validate, enrich, transform, package
    status: str  # pending, running, completed, failed
    progress_percent: int = 0
    message: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    summary: dict[str, Any] = Field(default_factory=dict)


class PipelineStatusResponse(BaseModel):
    """Schema for pipeline status response."""

    run_id: UUID
    project_id: UUID
    status: str  # pending, running, completed, failed, cancelled
    current_stage: Optional[str] = None
    progress_percent: int = 0
    stages: list[StageProgress] = Field(default_factory=list)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None


class PipelineRunResponse(BaseModel):
    """Schema for pipeline run response."""

    id: UUID
    project_id: UUID
    ifc_file_id: Optional[UUID] = None
    status: str
    current_stage: Optional[str] = None
    progress_percent: int = 0
    configuration: dict[str, Any] = Field(default_factory=dict)
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class PipelineHistoryResponse(BaseModel):
    """Schema for pipeline history response."""

    items: list[PipelineRunResponse]
    total: int
